### PR TITLE
additions to suggest telemetry

### PR DESCRIPTION
**Pull request by justschen**
ref https://github.com/microsoft/vscode/pull/225001

- we don't really care about the label anymore + it might be PII so we are removing
- we care more when there are multiple suggestions that have the same label - which one do users click? the first? second? third?
- this adds a separate telemetry event which captures the `index` of the selected suggestion, as well as the `firstIndex` which is the first instance that chosen label shows up. the `kind` may also be interesting to see (although we know the sort order prioritizes local first, etc, we can maybe see in the telemetry in cases the first one is not selected, what type _is_ selected)
- We only send this event when there are multiple suggestions with the same label as the selected one. 

example:
Screenshot 2024-08-13 at 11 24 51 PM
here if we select that Item, we log the first index as 2, but the chosen index as 3

TLDR: we want to see that when there are multiple labels that are the same, are people choosing the first one in the sorted list of completions

### CODE REVIEW

File: src/vs/editor/contrib/suggest/browser/suggestController.ts [hunk 1]
```diff
@@ -558,12 +560,54 @@ export class SuggestController implements IEditorContribution {
...
 			resolveDuration: item.resolveDuration,
 			commandDuration: commandExectionDuration,
-			additionalEditsAsync: additionalEditsAppliedAsync,
-			index,
-			label: item.textLabel
+			additionalEditsAsync: additionalEditsAppliedAsync
 		});
 	}
 
+	private _reportSuggestionIndexTelemetry(item: CompletionItem, index: number) {
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/browser/suggestController.ts:567]: IntelliSense happens *a lot* and without a hardcoded guard you'll be overloading the service. You should have some throttling, similar to `_reportSuggestionAcceptedTelemetry` or pull that throttle into the call site

File: src/vs/editor/contrib/suggest/browser/suggestController.ts [hunk 2]
```diff
@@ -257,6 +259,9 @@ export class SuggestController implements IEditorContribution {
 			if (index === -1) {
 				index = 0;
 			}
+
+			this.completionItems = e.completionModel.items;
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/browser/suggestController.ts:263]: We should not store `completionItems` in the controller. This is data of a particular invocation and not state. Also, the `_insertSuggestion` has already access to the model and all items

File: src/vs/editor/contrib/suggest/browser/suggestController.ts [hunk 3]
```diff
@@ -510,6 +515,7 @@ export class SuggestController implements IEditorContribution {
 		// clear only now - after all tasks are done
 		Promise.all(tasks).finally(() => {
 			this._reportSuggestionAcceptedTelemetry(item, model, isResolved, _commandExectionDuration, _additionalEditsAppliedAsync, event.index);
+			this._reportSuggestionIndexTelemetry(item, event.index);
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/browser/suggestController.ts:518]: use `event.model`...

File: src/vs/editor/contrib/suggest/browser/suggestController.ts [hunk 4]
```diff
@@ -558,12 +560,54 @@ export class SuggestController implements IEditorContribution {
...
+		// Determine if we have duplicates in the list
+		const hasOtherInstances = (items: CompletionItem[], label: string, index: number): boolean => {
+			return items.some((item, i) => i !== index && item.textLabel === label);
+		};
+
+		const hasDuplicates = hasOtherInstances(completionItems, selectedLabel, index);
+
+		// If duplicate, find first duplicate in the list
+		const firstIndex = hasDuplicates ? completionItems.findIndex((completion) => completion.textLabel === selectedLabel) : -1;
+
```
Comment (jrieken) [inline src/vs/editor/contrib/suggest/browser/suggestController.ts:586]: I think you improve performance here because an array is allocated (`splice`) and later that array is looped over two times. Can't you have a for-loop (till 30 or item length) that populates a map that is keyed by label and contains indexes? With that a simple look you give the first duplicate (if any).

Comment (jrieken) [reply src/vs/editor/contrib/suggest/browser/suggestController.ts:586]: Also note the danger of using (destructive) `splice` on a property. The property should not exist (see above) but splice would alter it and potentially trigger weird effects.

**Review by jrieken — CHANGES_REQUESTED** [review]
Thanks. Left a few comments, most importantly is the throttle because accepting intellisense roughly happens a few times with each millisecond that passes

**Review by jrieken — APPROVED** [review]
